sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -1216,7 +1216,7 @@ class Dataset[T] private[sql](
    */
   @scala.annotation.varargs
   def hint(name: String, parameters: Any*): Dataset[T] = withTypedPlan {
-    UnresolvedHint(name, parameters, logicalPlan)
+    UnresolvedHint(name, parameters, planWithBarrier)
```

Comment:
I think `ResolveBroadcastHints` rule will traverse recursively the children of logical plan. If we wrap it with a barrier, we can't be traverse down the tree.

Reply:
My understanding however is that `planWithBarrier` is already analyzed (and `ResolveBroadcastHints` as the very first rule had its chance to do its work). That's the extra processing `hint` does every time it's called. Using `planWithBarrier` makes it less "painful".

Just use `hint` twice and see the analyzed plan.

Reply:
I think @viirya has a valid concern. think about
```
val df1 = spark.table("t").select("id")
df1.hint("broadcast", "id")
```
We should transform down the plan of `df1`, find the bottom table relation and apply the hint.

Reply:
I thought that that's what `ResolveBroadcastHints` does --> https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveHints.scala#L93-L101, doesn't it? I'm going to write a test case for it to confirm (and that's what I was asking for in the email to dev@spark the other day).

Reply:
@jaceklaskowski Because the logical plan is wrapped in analysis barrier, `ResolveBroadcastHints` can't traverse down it to reach the `UnresolvedRelation` at https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveHints.scala#L60-L61.

Reply:
a possible workaround is to explicitly go through the barrier in the hint resolution rules, so that we can still use barrier here and skip analysis in other analyzer rules.

Reply:
If we do that, it is basically as the same as passing the logical plan without barrier.

Reply:
ah, I see. makes sense to me.

Reply:
shall we go with this workaround?